opt: eliminate LIMIT in right side of semi or anti join

mgartner opened the pull request:
#### opt: add tests for IntroduceExistsLimit normalization rule

This change adds tests for the IntroduceExistsLimit normalization rule,
which were previously not present. These tests should catch any
regressions of this rule.

Release note: None

#### opt: match all children node types in IntroduceExistsLimit

This change removes the ^(Project | GroupBy | DistinctOn) name match
pattern on the Exists input for the IntroduceExistsLimit rule.

Expression trees that match the Project, GroupBy, or DistinctOn patterns
already having the IntroduceExistsLimit rule applied after the
EliminateExistsProject and EliminateExistsGroupBy rules discard nodes.
Therefore this change simplifies the IntroduceExistsLimit rule without
changing the normalization process.

Release note: None

#### opt: eliminate LIMIT in right side of semi or anti join

A Limit expression with a positive limit is a no-op when used with a
SemiJoin or AntiJoin operator that has a True filter. By eliminating
this no-op Limit expression, the query can be decorrelated.

Fixes #45852

Release note (performance improvement): The query planner can now
decorrelate correlated exists subqueries with LIMIT expressions, leading
to better query plans.

mgartner committed:
opt: eliminate LIMIT in right side of semi or anti join

A Limit expression with a positive limit is a no-op when used with a
SemiJoin or AntiJoin operator that has a True filter. By eliminating
this no-op Limit expression, the query can be decorrelated.

Fixes #45852

Release note (performance improvement): The query planner can now
decorrelate correlated exists subqueries with LIMIT expressions, leading
to better query plans.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46540)
<!-- Reviewable:end -->

mgartner commented on pkg/sql/opt/norm/rules/limit.opt:
```diff
@@ -14,6 +14,19 @@
+# EliminateLimitInExists discards a Limit operator with a positive limit inside
+# an Exist operator.
+[EliminateLimitInExists, Normalize]
+(Exists
```
Should this be more general? E.g. should it match SemiJoin and AntiJoin patterns instead of only matching exists specifically?

mgartner replied:
This also doesn't restrict the normalization to only applying inside a `WHERE` clause, which maybe isn't desired? For example, should the limit not be eliminated in something like:

```
SELECT EXISTS(SELECT * FROM (VALUES ('radu', 'andy')) LIMIT 1)
```

mgartner commented on pkg/sql/opt/norm/testdata/rules/limit:
```diff
@@ -109,6 +109,61 @@ limit
+
+# Don't eliminate a non-positive limit.
+norm
+SELECT * FROM a WHERE EXISTS (SELECT * FROM b where b.y = a.k LIMIT 0)
```
I'm not sure this test makes sense. On one hand I want EliminateLimitInExists to not affect the expression tree when the limit is non-positive. But the LIMIT expressions ends up being eliminated anyway by another normalization. Thoughts?

andy-kimball reviewed (commented):
Looks like you've picked up quickly on how to write one of these.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3Dy56k-3igZYKLy_x1:bfkmwa1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/norm/rules/limit.opt, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3DyOQ47m31Bxq-ypzz:-M3DyOQ47m31Bxq-yq--:bh13jmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f597fe59137b37b12d732578ca703e01fe6424b/pkg/sql/opt/norm/rules/limit.opt#L18)):*
> ```
> 
> # EliminateLimitInExists discards a Limit operator with a positive limit inside
> # an Exist operator.
> ```

NIT: Maybe expand this description to include the "why" of the rule - that the `Limit` gets in the way of decorrelation rules.
___
*[pkg/sql/opt/norm/rules/limit.opt, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3DuW3E1choHGdaLmWN-r1-20:-M3DwJ_qBoanIgEgujiX:bfk30e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f597fe59137b37b12d732578ca703e01fe6424b/pkg/sql/opt/norm/rules/limit.opt#L20)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Should this be more general? E.g. should it match SemiJoin and AntiJoin patterns instead of only matching `EXISTS` specifically?
</blockquote></details>

I think this is OK, with one change: we should only match this pattern when `(HasOuterCols $input)`. See the `IntroduceExistsLimit` rule in `sql/opt/norm/rules/scalar.opt`, which will introduce a `Limit` in the uncorrelated case. If we don't check `HasOuterCols` here, then this rule would then just remove that.

I think Raphael never added tests when he added the `IntroduceExistsLimit`, so that explains why you didn't see a test case fail due to this conflict. Can you add a test case there? In fact, I'd move this rule there. For `opt` files that are named after operators (like `limit.opt`, 'select.opt`, etc), we group rules by the top-level operator they match. So `EliminateLimitInExists` should be located right alongside `IntroduceExistsLimit` (and probably renamed to `EliminateExistsLimit`.
___
*[pkg/sql/opt/norm/testdata/rules/limit, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3DuW3E1choHGdaLmWO-r1-149:-M3DxeWHEiPr5b9nGcCQ:bf3m56z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f597fe59137b37b12d732578ca703e01fe6424b/pkg/sql/opt/norm/testdata/rules/limit#L149)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not sure this test makes sense. On one hand I want to ensure EliminateLimitInExists does not affect the expression tree when the limit is non-positive. But in this example the LIMIT expression ends up being eliminated by another normalization. Thoughts?
</blockquote></details>

Use the `norm expect-not=EliminateLimitInExists` directive. That will ensure that the rule doesn't fire.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/norm/rules/limit.opt, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3DuW3E1choHGdaLmWN-r1-20:-M3DzjoxAcL7hxrF8NDT:b-wzogv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f597fe59137b37b12d732578ca703e01fe6424b/pkg/sql/opt/norm/rules/limit.opt#L20)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This also doesn't restrict the normalization to only applying inside a `WHERE` clause, which maybe isn't desired? For example, should the limit not be eliminated in something like:

```
SELECT EXISTS(SELECT * FROM (VALUES ('radu', 'andy')) LIMIT 1)
```
</blockquote></details>

That'd be an uncorrelated case, so this rule would not fire once you add `(HasOuterCols $input)`. We only want this rule to fire when there are one or more "outer columns", which are columns that refer to tables outside the scope of the `EXISTS`.


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: add tests for IntroduceExistsLimit normalization rule

This change adds tests for the IntroduceExistsLimit normalization rule,
which were previously not present. These tests should catch any
regressions of this rule.

Release note: None

mgartner committed:
opt: match all children node types in IntroduceExistsLimit

This change removes the ^(Project | GroupBy | DistinctOn) name match
pattern on the Exists input for the IntroduceExistsLimit rule.

Expression trees that match the Project, GroupBy, or DistinctOn patterns
already having the IntroduceExistsLimit rule applied after the
EliminateExistsProject and EliminateExistsGroupBy rules discard nodes.
Therefore this change simplifies the IntroduceExistsLimit rule without
changing the normalization process.

Release note: None

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, very nice. Just wait to merge until we cut the release branch. Should be soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3IcfXuCmA4oB2SXnPp:b-3bvqw5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3IdN4Z0OTZLCekEzUN:bnahwna)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/norm/testdata/rules/scalar, line 383 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3HwGp62hTccmcHRbn-:-M3HwGp62hTccmcHRbn0:b2kzgtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/584aa8b65e8f32dcec142fcd82a18aa699b80971/pkg/sql/opt/norm/testdata/rules/scalar#L383)):*
> ```
> # IntroduceExistsLimit
> # --------------------------------------------------
> norm expect=(IntroduceExistsLimit)
> ```

[nit] no paren necessary with a single rule (same below)


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nicely done!

I think you should also add a release note to the last commit since this will affect customers (in a positive way).

> Reviewed 2 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3MNNdv1kfOchXhvhFM:b-oeqqtf)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3MLWYJ6akJ-RUMn9dZ:-M3MLWYJ6akJ-RUMn9d_:b1nt3jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a33e5b30589036cd663e0174bd03f39669e0df6/pkg/sql/opt/norm/rules/scalar.opt#L115)):*
> ```
> # and it would recurse otherwise.
> #
> # We avoid this rule if the query is decorrelated because the
> ```

while you're editing this file.... I think "decorrelated" should be changed to "correlated" here
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3MLjWy8PFFM-S90C6b:-M3MLjWy8PFFM-S90C6c:blbr8p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a33e5b30589036cd663e0174bd03f39669e0df6/pkg/sql/opt/norm/rules/scalar.opt#L119)):*
> ```
> # (It will be worth re-considering this when a general-purpose apply
> # operator is supported - in that case it can be definitely worthwhile
> # pushing down a LIMIT 1 to limit the amount of work done on every row.)
> ```

Also we have a general-purpose apply operator now.... is this worth reconsidering? @andy-kimball @raduberinde 


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3MLjWy8PFFM-S90C6b:-M3OSGty1qtbkyOKSD9Z:b27ykq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a33e5b30589036cd663e0174bd03f39669e0df6/pkg/sql/opt/norm/rules/scalar.opt#L119)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Also we have a general-purpose apply operator now.... is this worth reconsidering? @andy-kimball @raduberinde 
</blockquote></details>

No, I don't think the apply operator makes a difference. However, the soft-limit physical property may make this rule obsolete, as long as `Exists` is wired up to use it.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3MLjWy8PFFM-S90C6b:-M3go4iu1e6KUuuQ9ajo:badik3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a33e5b30589036cd663e0174bd03f39669e0df6/pkg/sql/opt/norm/rules/scalar.opt#L119)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

No, I don't think the apply operator makes a difference. However, the soft-limit physical property may make this rule obsolete, as long as `Exists` is wired up to use it.
</blockquote></details>

I don't think `Exists` is wired up to use soft limits, but thinking about this more, I don't think this really matters since we always convert correlated `Exists` to semi or anti join. Using a hard limit (as we do in this rule) for the non-correlated case seems better than a soft limit, no?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3MLWYJ6akJ-RUMn9dZ:-M3hR1YpCOwFmR-Kve2T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a33e5b30589036cd663e0174bd03f39669e0df6/pkg/sql/opt/norm/rules/scalar.opt#L115)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

while you're editing this file.... I think "decorrelated" should be changed to "correlated" here
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Release note (performance improvement): The query planner can now decorrelate correlated subqueries...

[nit] I'd say "some correlated subqueries" or "correlated EXISTS subqueries"

> Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3hS-nr2Nh-97F7YnnR:b-4nhn2k)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hSHE24cI4hvIK0VPm:-M3hSHE24cI4hvIK0VPn:bp428rz) ([raw file](https://github.com/cockroachdb/cockroach/blob/30abf53d22e7f75c0a1e863544c32349d68316b8/pkg/sql/opt/norm/rules/scalar.opt#L138)):*
> ```
> # can be decorrelated into a more efficient SemiJoin or AntiJoin.
> #
> # Note that this rule uses HasOutCols to ensure that it only matches correlated
> ```

HasOutCols -> HasOuterCols
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hSSTp3FibP_vEAH-j:-M3hSSTp3FibP_vEAH-k:b-mk0njb) ([raw file](https://github.com/cockroachdb/cockroach/blob/30abf53d22e7f75c0a1e863544c32349d68316b8/pkg/sql/opt/norm/rules/scalar.opt#L140)):*
> ```
> # Note that this rule uses HasOutCols to ensure that it only matches correlated
> # subqueries. There is no need to discard limits from non-correlated subqueries.
> # Limits are preferred for non-correlated subqueries. See IntroduceExistsLimit
> ```

[nit] non-correlated subqueries -> non-correlated EXISTS subqueries


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 138 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hSHE24cI4hvIK0VPm:-M3hTx7n4HAgfn-u0r3N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30abf53d22e7f75c0a1e863544c32349d68316b8/pkg/sql/opt/norm/rules/scalar.opt#L138)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

HasOutCols -> HasOuterCols
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hSSTp3FibP_vEAH-j:-M3hTxN-4pbZXOKAFj1p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30abf53d22e7f75c0a1e863544c32349d68316b8/pkg/sql/opt/norm/rules/scalar.opt#L140)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] non-correlated subqueries -> non-correlated EXISTS subqueries
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3hklatDKvDSm72jUgb:brao51c) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
> ```
> SELECT * FROM (SELECT * FROM xysd LIMIT 10) WHERE EXISTS(SELECT * FROM uv WHERE x=u LIMIT 5)
> ----
> semi-join-apply
> ```

I think this might be a case where this test no longer is testing what it is supposed to. Based on the comment, I think this is specific to semi-join-apply only. I think if I make the EXISTS subquery non-correlated, it will remain a semi-join-apply, but I'm not sure if that's ok for the "calcualte ... cardinality" part of this test case.
___
*[pkg/sql/opt/memo/testdata/logprops/join, line 719 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hlExJBvANpScldSW-:-M3hlExJBvANpScldSW0:b93y50c) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L719)):*
> ```
> SELECT * FROM (SELECT * FROM xysd LIMIT 10) WHERE NOT EXISTS(SELECT * FROM uv WHERE x=u LIMIT 5)
> ----
> anti-join-apply
> ```

Ditto on this one. This test case should remain an anti-join-apply.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3htKZB2bczImYm0zZ5:bhmmdbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think this might be a case where this test no longer is testing what it is supposed to. Based on the comment, I think this is specific to semi-join-apply only. I think if I make the EXISTS subquery non-correlated, it will remain a semi-join-apply, but I'm not sure if that's ok for the "calcualte ... cardinality" part of this test case.
</blockquote></details>

You're right. It's getting harder and harder to find expressions that the optimizer can't decorrelate. Maybe we should switch over to use the `expr` command. Here's an example in `pkg/sql/opt/memo/testdata/stats/join`:
```
expr format=show-all
(MakeLookupJoin
  (Scan [ (Table "abc") (Cols "a,b,c") ])
  [ (JoinType "inner-join") (Table "def") (Index "def@primary") (KeyCols "a,b") (Cols "a,b,c,d,f,e") ]
  [ ]
)
```
This command lets us directly create exactly the opt tree that we want. Might take some digging to find out exactly how to translate this expression to use it, though.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @andy-kimball, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3lWxfz62PuhPC8ss6q:b-mzit52) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You're right. It's getting harder and harder to find expressions that the optimizer can't decorrelate. Maybe we should switch over to use the `expr` command. Here's an example in `pkg/sql/opt/memo/testdata/stats/join`:
```
expr format=show-all
(MakeLookupJoin
  (Scan [ (Table "abc") (Cols "a,b,c") ])
  [ (JoinType "inner-join") (Table "def") (Index "def@primary") (KeyCols "a,b") (Cols "a,b,c,d,f,e") ]
  [ ]
)
```
This command lets us directly create exactly the opt tree that we want. Might take some digging to find out exactly how to translate this expression to use it, though.
</blockquote></details>

Great I'll try that. So this test case must have a correlated subquery then? Why?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3ldQzB719UpTNSPe0E:bp37fs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Great I'll try that. So this test case must have a correlated subquery then? Why?
</blockquote></details>

The apply variants are only used when the right side of the join is correlated with the left side (it is executed as a loop - for every row on the left side, execute the entire right side).

But for this case, if you use `expr`, it doesn't matter. This code only cares about testing the properties of the top-level node. You can even use`FakeRel` to directly set the properties of the inputs. https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/memo/testdata/stats/select#L1245


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3m1RLt9qH3_UYWlp3A:b2fyu9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The apply variants are only used when the right side of the join is correlated with the left side (it is executed as a loop - for every row on the left side, execute the entire right side).

But for this case, if you use `expr`, it doesn't matter. This code only cares about testing the properties of the top-level node. You can even use`FakeRel` to directly set the properties of the inputs. https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/memo/testdata/stats/select#L1245
</blockquote></details>

Ok I gave the `expr` with a `FakeRel` a shot. Is this what you had in mind? I think this is a valid test to ensure that the Cardinality of the left side of the semi-join-apply (0-10) is propagated up to the semi-join-apply node.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  (don't forget to update the release note)

> Reviewed 4 of 5 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3m7iTECZjiYcL0EDCM:b-vlt94f)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3mJ97fAe1ioEodX9UV:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/memo/testdata/logprops/join, line 657 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-M3hklatDKvDSm72jUga:-M3mJ3Ne5NiFictfhB3t:birp450) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cb2eb7cac9a8e918721b5f5cbdb0f4a0823f45/pkg/sql/opt/memo/testdata/logprops/join#L657)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Ok I gave the `expr` with a `FakeRel` a shot. Is this what you had in mind? I think this is a valid test to ensure that the Cardinality of the left side of the semi-join-apply (0-10) is propagated up to the semi-join-apply node.
</blockquote></details>

Yeah, looks good to me. Thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46540#-:-M3sJ_T11Vr_WRgYiv8M:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

mgartner commented:
bors r+
